build

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env bash
+
+# build - build Jellyfin binaries or packages
+
+set -o errexit
+set -o pipefail
+
+# The list of possible package actions (except 'clean')
+actions=( 'build' 'package' 'sign' 'publish' )
+
+# The list of possible platforms, based on directories under 'deployment/'
+platforms=( $(
+    find deployment/ -maxdepth 1 -mindepth 1 -type d | sed 's/deployment\///' | sort
+) )
+
+# The list of standard dependencies required by all build scripts; individual
+# action scripts may specify their own dependencies
+dependencies=( 'tar' 'zip' )
+
+usage() {
+    echo -e "build - build Jellyfin binaries or packages"
+    echo -e ""
+    echo -e "Usage:"
+    echo -e " $ build --list-platforms"
+    echo -e " $ build --list-actions <platform>"
+    echo -e " $ build [-b/--web-branch <web_branch>] <platform> <action>"
+    echo -e ""
+    echo -e "The web_branch defaults to 'master'."
+    echo -e "To build all platforms, use 'all'."
+    echo -e "To perform all build actions, use 'all'."
+    echo -e "Build output files are collected at '../jellyfin-build/<platform>'."
+}
+
+if [[ -z $1 ]]; then
+    usage
+    exit 1
+fi
+
+# Verify required utilities are installed
+for utility in ${dependencies[@]}; do
+    if ! which ${utility} &>/dev/null; then
+        echo -e "This script requires the '${utility}' utility."
+    fi
+done
+
+# List all available platforms then exit
+if [[ $1 == '--list-platforms' ]]; then
+    echo -e "Available platforms:"
+    for platform in ${platforms[@]}; do
+        echo -e "  ${platform}"
+    done
+    exit 0
+fi
+
+# List all available actions for a given platform then exit
+if [[ $1 == '--list-actions' ]]; then
+    platform="$2"
+    if [[ ! " ${platforms[@]} " =~ " ${platform} " ]]; then
+        echo "Platform $platform does not exist."
+        exit 1
+    fi
+    echo -e "Available actions for platform ${platform}:"
+    for action in ${actions[@]}; do
+        if [[ -f deployment/${platform}/${action}.sh ]]; then
+            echo -e "  ${action}"
+        fi
+    done
+    exit 0
+fi
+
+# Initialize submodules
+git submodule update --init --recursive
+
+# Parse branch option
+if [[ $1 == '-b' || $1 == '--web-branch' ]]; then
+	web_branch="$2"
+	shift 2
+else
+    web_branch="master"
+fi
+# configure branch
+pushd MediaBrowser.WebDashboard/jellyfin-web
+git checkout ${web_branch} || {
+    echo "jellyfin-web branch ${web_branch} is invalid."
+    exit 1
+}
+git pull --rebase
```

Comment:
cannot you just do `git checkout "origin/${web_branch}"`?

Reply:
Will that always pull down the commits from the last run? If so that'll work.

Reply:
You will need a "git fetch origin". And it leaves the repo in a detached head state (possibly).

Reply:
That shouldn't matter too much, as long as the code is at the right HEAD. So something like:

```
git fetch --all
git checkout origin/${web_branch}
```

Though do we even need the "origin" ref? I'll do some testing with this.

Reply:
OK so this gives `git:(HEAD detached at origin/dev)`, but I think that's totally fine. Could use `-b` too but not sure whether to even bother since it's a submodule.

Reply:
Actually yea this is ideal, it resets every time so its always at the latest for that branch, while not cluttering up the submodule with local branches.

```
Submodule path 'MediaBrowser.WebDashboard/jellyfin-web': checked out '4678528d0028685b45c7c6daa2e24b72a363535b'
~/Projects/Jellyfin/jellyfin/MediaBrowser.WebDashboard/jellyfin-web ~/Projects/Jellyfin/jellyfin
Fetching origin
Previous HEAD position was 4678528d0... First separation commit.
HEAD is now at 7417b844f... Merge pull request #27 from hawken93/fix_503
```

Reply:
Well this might leave the repo in a non-desirable state, since any and all commits will be lost.

Reply:
No one should be working in the submodule directly :-/ That's just a terrible practice.

Reply:
That can work, just as a fail-save to say "Hey, fix this, the submodule should be clean".

Reply:
Done, it now exits with an error (and a reminder not to do this...) if it finds any uncleanliness in the working tree of the submodule.